man/nxsdata.1
.\" Man page generated from reStructuredText.
.
.
.nr rst2man-indent-level 0
.
.de1 rstReportMargin
\\$1 \\n[an-margin]
level \\n[rst2man-indent-level]
level margin: \\n[rst2man-indent\\n[rst2man-indent-level]]
-
\\n[rst2man-indent0]
\\n[rst2man-indent1]
\\n[rst2man-indent2]
..
.de1 INDENT
.\" .rstReportMargin pre:
. RS \\$1
. nr rst2man-indent\\n[rst2man-indent-level] \\n[an-margin]
. nr rst2man-indent-level +1
.\" .rstReportMargin post:
..
.de UNINDENT
. RE
.\" indent \\n[an-margin]
.\" old: \\n[rst2man-indent\\n[rst2man-indent-level]]
.nr rst2man-indent-level -1
.\" new: \\n[rst2man-indent\\n[rst2man-indent-level]]
.in \\n[rst2man-indent\\n[rst2man-indent-level]]u
..
.TH "NXSDATA" "1" "Nov 08, 2024" "4.18" "NXSTools"
.SH NAME
nxsdata \- run NeXus Data Writer
.SH DESCRIPTION
.sp
The nxsdata program is a command\-line interface to Nexus Data Tango Server.
Program allows one to store NeXuS Data in H5 files.
The writer provides storing data from other Tango devices, various databases
as well as passed by a user client via JSON strings.
.SH SYNOPSIS
.INDENT 0.0
.INDENT 3.5
.sp
.nf
.ft C
nxsdata <command> [\-s <nexus_server>]  [<arg1> [<arg2>  ...]]
.ft P
.fi
.UNINDENT
.UNINDENT
.INDENT 0.0
.TP
.B Commands:
.INDENT 7.0
.TP
.B openfile [\-s <nexus_server>]  <file_name>
open new H5 file
.TP
.B setdata [\-s <nexus_server>] <json_data_string>
assign global JSON data
.TP
.B openentry [\-s <nexus_server>] <xml_config>
create new entry
.TP
.B record [\-s <nexus_server>]  <json_data_string>
record one step with step JSON data
.TP
.B closeentry [\-s <nexus_server>]
close the current entry
.TP
.B closefile [\-s <nexus_server>]
close the current file
.TP
.B servers [\-s <nexus_server/host>]
get lists of tango data servers from the current tango host
.UNINDENT
.TP
.B Options:
.INDENT 7.0
.TP
.B  \-h\fP,\fB  \-\-help
show this help message and exit
.TP
.BI \-s \ SERVER\fR,\fB \ \-\-server\fB= SERVER
tango data server device name
.UNINDENT
.UNINDENT
.SH EXAMPLE
.INDENT 0.0
.INDENT 3.5
.sp
.nf
.ft C
nxsdata openfile \-s p02/tangodataserver/exp.01  /user/data/myfile.h5
.ft P
.fi
.UNINDENT
.UNINDENT
.SH AUTHOR
Jan Kotanski
.SH COPYRIGHT
2012-2018 DESY, Jan Kotanski <jkotan@mail.desy.de>

GNU GENERAL PUBLIC LICENSE, version 3
.\" Generated by docutils manpage writer.
.


ChangeLog
2024-11-08 Jan Kotanski <jankotan@gmail.com>
	* fix typo in h5rediswriter (#729)
	* add `session` parameter for h5redis Writer property (#731)
	* tagged as v4.18.0

2024-11-06 Jan Kotanski <jankotan@gmail.com>
	* add support for append end extend actions in copy-map files (#726)
	* add singlelist type in group-map (#726)
	* tagged as v4.17.0

2024-10-22 Jan Kotanski <jankotan@gmail.com>
	* improve datasetName values generation in the ScanNames mode (#723)
	* tagged as v4.16.2

2024-09-27 Jan Kotanski <jankotan@gmail.com>
	* fix for metadata parser of groups without NX_class (#718)
	* fix minimal_value and maximal_value to minimum_value and maximum_value NXlog attributes (#718)
	* tagged as v4.16.1

2024-09-26 Jan Kotanski <jankotan@gmail.com>
	* add enum as a valid secop type(#715)
	* tagged as v4.16.0

2024-09-11 Jan Kotanski <jankotan@gmail.com>
	* add support for link, key and function for secop meaning dictionary (#712)
	* tagged as v4.15.0

2024-08-08 Jan Kotanski <jankotan@gmail.com>
	* add support for secop meaning dictionary (#709)
	* tagged as v4.14.0

2024-08-08 Jan Kotanski <jankotan@gmail.com>
	* fix typo in for device_name in nxsetup (#706)
	* tagged as v4.13.1

2024-07-19 Jan Kotanski <jankotan@gmail.com>
	* improve findAttribute by adding moduleAttributeMap (#696)
	* add ScanNamesNoMetadata and ScanNamesNoGrouping variables (#702)
	* tagged as v4.13.0

2024-07-15 Jan Kotanski <jankotan@gmail.com>
	* create output dir is does not exists (#696)
	* fix ScanFileinScanNameDir name to ScanNames (#694)
	* fix support for ScanNames=False (#695)
	* tagged as v4.12.0

2024-07-11 Jan Kotanski <jankotan@gmail.com>
	* add --online-xml-file option for nxscreate stdcomp (#687)
	* add support for ScanFileinScanNameDir mode for detector template (#689)
	* fix for eigerdectris VDS (#691)
	* tagged as v4.11.0

2024-07-10 Jan Kotanski <jankotan@gmail.com>
	* add support ScanFileInScanNameDir env variable
	* tagged as v4.10.0

2024-07-10 Jan Kotanski <jankotan@gmail.com>
	* improve support for eiger RoiMode attribute
	* tagged as v4.9.1

2024-07-09 Jan Kotanski <jankotan@gmail.com>
	* add support for eiger RoiMode attribute
	* tagged as v4.9.0

2024-07-07 Jan Kotanski <jankotan@gmail.com>
	* hide STEP datasources in  eiger templates (#673)
	* hide STEP datasources in  lambdavds templates (#675)
	* tagged as v4.8.0

2024-07-03 Jan Kotanski <jankotan@gmail.com>
	* add --id-format to nxsfileinfo metadata (#669)
	* tagged as v4.7.0

2024-07-03 Jan Kotanski <jankotan@gmail.com>
	* fix wrong parsing of summer time  (#666)
	* tagged as v4.6.2

2024-07-02 Jan Kotanski <jankotan@gmail.com>
	* fix pilc file name (#663)
	* tagged as v4.6.1

2024-07-02 Jan Kotanski <jankotan@gmail.com>
	* add pilc component template (#661)
	* tagged as v4.6.0

2024-06-27 Jan Kotanski <jankotan@gmail.com>
	* improve fix for enlarge waiting time for the server in nxsetup (#659)
	* tagged as v4.5.1

2024-06-27 Jan Kotanski <jankotan@gmail.com>
	* enlarge waiting time for the server in nxsetup (#657)
	* tagged as v4.5.0

2024-06-11 Jan Kotanski <jankotan@gmail.com>
	* add eiger9m16vds template (#654)
	* add eiger4m16vds template (#654)
	* add eiger1m16vds template (#654)
	* add eiger4m32vds template (#654)
	* add eiger1m32vds template (#654)
	* tagged as v4.4.0

2024-06-10 Jan Kotanski <jankotan@gmail.com>
	* add eigerdectrismesh template (#649)
	* add eiger9m32vds template (#652)
	* fix typos in stepindex datasources
	* tagged as v4.3.0

2024-06-05 Jan Kotanski <jankotan@gmail.com>
	* reingest scan metadata if it is repeated with the same ScanID (#647)
	* tagged as v4.2.1

2024-06-03 Jan Kotanski <jankotan@gmail.com>
	* adapt append_scicat_dataset to _%05d and _{ScanID:05d} formatting (#643)
	* tagged as v4.2.0

2024-05-23 Jan Kotanski <jankotan@gmail.com>
	* replace deprecated tz.localize() with datetime.replace() (#640)
	* tagged as v4.1.1

2024-05-22 Jan Kotanski <jankotan@gmail.com>
	* adapt the entry name in the appendentry mode to NXS_FileRecorder v3.21.0 (#637)
	* tagged as v4.1.0

2024-05-15 Jan Kotanski <jankotan@gmail.com>
	* reduce number of plots in h5redis writer (#635)
	* tagged as v4.0.2

2024-05-12 Jan Kotanski <jankotan@gmail.com>
	* add finish for h5redis writer
	* tagged as v4.0.1

024-05-10 Jan Kotanski <jankotan@gmail.com>
	* add support for h5redis writer
	* drop support for python2
	* tagged as v4.0.0

2024-03-27 Jan Kotanski <jankotan@gmail.com>
	* add nxsetup wait subcommand (#624)
	* add --timeout -no-wait for  nonxsetup subcommands (#624)
	* extend nxsetup set to create a general device (#624)
	* tagged as v3.67.0

2024-03-22 Jan Kotanski <jankotan@gmail.com>
	* add --confiserver option to nxsetup set
	* tagged as v3.66.0

2024-03-14 Jan Kotanski <jankotan@gmail.com>
	* increase waiting time for restarting servers (#618)
	* tagged as v3.65.1

2024-02-22 Jan Kotanski <jankotan@gmail.com>
	* spilt NXsample to sample and sample_environment like for SECoP components (#615)
	* remove --strict and  --node-in-instrument (#615)
	* add --sample-environment (#615)
	* tagged as v3.65.0

2024-02-19 Jan Kotanski <jankotan@gmail.com>
	* store importance as a measurement attribute (#608)
	* add dynamic secop helper links (#610)
	* add sample nxdata from secop (#612)
	* tagged as v3.64.0

2024-02-09 Jan Kotanski <jankotan@gmail.com>
	* add techniques only with the pid (#605)
	* tagged as v3.63.1

2024-02-08 Jan Kotanski <jankotan@gmail.com>
	* add first, last, firstlast, endpoints type in group-map (#602)
	* tagged as v3.63.0

2024-01-24 Jan Kotanski <jankotan@gmail.com>
	* change --node-in-sample to  NOT --node-in-instrument (#599)
	* tagged as v3.62.0

2024-01-19 Jan Kotanski <jankotan@gmail.com>
	* convert nan and inf to null in (group)metadata (#594)
	* tagged as v3.61.1

2024-01-15 Jan Kotanski <jankotan@gmail.com>
	* fix  secop rotation_z coordiantes to [0,1,0] (#585)
	* add NXsensor description to secop components (#587)
	* add min and max type in group-map (#589)
	* tagged as v3.61.0


nxstools/nxsdata.py
#!/usr/bin/env python
#   This file is part of nexdatas - Tango Server for NeXus data writer
#
#    Copyright (C) 2012-2018 DESY, Jan Kotanski <jkotan@mail.desy.de>
#
#    nexdatas is free software: you can redistribute it and/or modify
#    it under the terms of the GNU General Public License as published by
#    the Free Software Foundation, either version 3 of the License, or
#    (at your option) any later version.
#
#    nexdatas is distributed in the hope that it will be useful,
#    but WITHOUT ANY WARRANTY; without even the implied warranty of
#    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
#    GNU General Public License for more details.
#
#    You should have received a copy of the GNU General Public License
#    along with nexdatas.  If not, see <http://www.gnu.org/licenses/>.
#

""" Command-line tool to ascess to Tango Data Server"""

import sys

import argparse

from .nxsdevicetools import (checkServer, listServers, openServer)
from .nxsargparser import (Runner, NXSArgParser, ErrorException)


class NexusServer(object):

    """ configuration server adapter
    """

    def __init__(self, device):
        """ constructor

        :param device: device name of configuration server
        :type device: :obj:`str`
        """
        #: (:class:`PyTango.DeviceProxy`) NeXus writer device proxy
        self.tdwServer = openServer(device)

    def openFile(self, filename):
        """ opens the h5 file

        :param filename: h5 file name
        :type filename: :obj:`str`
        """
        self.tdwServer.Init()
        self.tdwServer.FileName = str(filename)
        self.tdwServer.OpenFile()

    def setData(self, jsondata):
        """ sets the global JSON data

        :param jsondata: global JSON data
        :type jsondata: :obj:`str`
        """
        self.tdwServer.JSONRecord = str(jsondata)

    def openEntry(self, xmlconfig):
        """ opens an entry

        :param xmlconfig: xml configuration string
        :type xmlconfig: :obj:`str`
        """
        self.tdwServer.XMLSettings = str(xmlconfig)
        self.tdwServer.OpenEntry()

    def record(self, jsondata):
        """ records one step

        :param jsondata: step JSON data
        :type jsondata: :obj:`str`
        """
        self.tdwServer.Record(jsondata)

    def closeEntry(self):
        """ closes the entry
        """
        self.tdwServer.CloseEntry()

    def closeFile(self):
        """ closes the file

        """
        self.tdwServer.CloseFile()


class OpenFile(Runner):

    """ OpenFile runner"""

    #: (:obj:`str`) command description
    description = "open a new H5 file"
    epilog = "" \
        + " examples:\n" \
        + "       nxsdata openfile /tmp/watertest.nxs \n" \
        + "       nxsdata openfile -s p02/tangodataserver/exp.01  " \
        + "/user/data/myfile.h5\n" \
        + "\n"

    def create(self):
        """ creates parser

        """
        parser = self._parser
        parser.add_argument(
            "-s", "--server", dest="server",
            help="writer server device name")

    def postauto(self):
        """ parser creator after autocomplete run """
        parser = self._parser
        parser.add_argument(
            'args', metavar='file_name', type=str, nargs='?',
            help='new newxus file name')

    def run(self, options):
        """ the main program function

        :param options: parser options
        :type options: :class:`argparse.Namespace`
        :returns: output information
        :rtype: :obj:`str`
        """
        tdwserver = NexusServer(options.server)
        if not options.args or len(options.args) < 1:
            self._parser.print_help()
            sys.exit(255)
        return tdwserver.openFile(options.args[0])


class SetData(Runner):

    """ SetData runner"""

    #: (:obj:`str`) command description
    description = "assign global JSON data"
    epilog = "" \
        + " examples:\n" \
        + "       nxsdata setdata ... \n" \
        + "\n"

    def create(self):
        """ creates parser

        """
        parser = self._parser
        parser.add_argument(
            "-s", "--server", dest="server",
            help="writer server device name"
        )
        parser.add_argument(
            'args', metavar='json_data_string', type=str, nargs='?',
            help='json data string')

    def run(self, options):
        """ the main program function

        :param options: parser options
        :type options: :class:`argparse.Namespace`
        :returns: output information
        :rtype: :obj:`str`
        """
        tdwserver = NexusServer(options.server)
        if not options.args or len(options.args) < 1:
            self._parser.print_help()
            sys.exit(255)
        return tdwserver.setData(options.args[0].strip())


class OpenEntry(Runner):

    """ OpenEntry runner"""

    #: (:obj:`str`) command description
    description = "create new entry"
    epilog = "" \
        + " examples:\n" \
        + "       nxsdata openentry ... \n" \
        + "\n"

    def create(self):
        """ creates parser

        """
        parser = self._parser
        parser.add_argument(
            "-s", "--server", dest="server",
            help="writer server device name")
        parser.add_argument(
            'args', metavar='xml_config', type=str, nargs='?',
            help='nexus writer configuration string')

    def run(self, options):
        """ the main program function

        :param options: parser options
        :type options: :class:`argparse.Namespace`
        :returns: output information
        :rtype: :obj:`str`
        """
        tdwserver = NexusServer(options.server)
        if not options.args or len(options.args) < 1:
            self._parser.print_help()
            sys.exit(255)
        return tdwserver.openEntry(options.args[0].strip())


class Record(Runner):

    """ Record runner"""

    #: (:obj:`str`) command description
    description = "record one step with step JSON data"
    epilog = "" \
        + " examples:\n" \
        + "       nxsdata record ... \n" \
        + "\n"

    def create(self):
        """ creates parser

        """
        parser = self._parser
        parser.add_argument(
            "-s", "--server", dest="server",
            help="writer server device name")
        parser.add_argument(
            'args', metavar='json_data_string', type=str, nargs='?',
            help='json data string')

    def run(self, options):
        """ the main program function

        :param options: parser options
        :type options: :class:`argparse.Namespace`
        :returns: output information
        :rtype: :obj:`str`
        """
        tdwserver = NexusServer(options.server)
        return tdwserver.record(
            options.args[0].strip() if options.args else '{}')


class CloseEntry(Runner):

    """ CloseEntry runner"""

    #: (:obj:`str`) command description
    description = "close the current entry"
    epilog = "" \
        + " examples:\n" \
        + "       nxsdata closeentry \n" \
        + "\n"

    def create(self):
        """ creates parser

        """
        parser = self._parser
        parser.add_argument(
            "-s", "--server", dest="server",
            help="writer server device name"
        )

    def run(self, options):
        """ the main program function

        :param options: parser options
        :type options: :class:`argparse.Namespace`
        :returns: output information
        :rtype: :obj:`str`
        """
        tdwserver = NexusServer(options.server)
        return tdwserver.closeEntry()


class CloseFile(Runner):

    """ CloseFile runner"""

    #: (:obj:`str`) command description
    description = "close the current file"
    epilog = "" \
        + " examples:\n" \
        + "       nxsdata closefile \n" \
        + "\n"

    def create(self):
        """ creates parser

        """
        parser = self._parser
        parser.add_argument(
            "-s", "--server", dest="server",
            help="writer server device name"
        )

    def run(self, options):
        """ the main program function

        :param options: parser options
        :type options: :class:`argparse.Namespace`
        :returns: output information
        :rtype: :obj:`str`
        """
        tdwserver = NexusServer(options.server)
        return tdwserver.closeFile()


class Servers(Runner):

    """ Servers runner"""

    #: (:obj:`str`) command description
    description = "get lists of tango data servers from " \
                  + "the current tango host"
    epilog = "" \
        + " examples:\n" \
        + "       nxsdata servers \n" \
        + "\n"

    def create(self):
        """ creates parser

        """
        parser = self._parser
        parser.add_argument(
            "-s", "--server", dest="server",
            help="tango host or writer server device name"
        )

    def run(self, options):
        """ the main program function

        :param options: parser options
        :type options: :class:`argparse.Namespace`
        :returns: output information
        :rtype: :obj:`str`
        """
        return "\n".join(listServers(options.server, 'NXSConfigServer'))


def main():
    """ the main program function
    """

    #: pipe arguments
    pipe = ""
    if not sys.stdin.isatty():
        pp = sys.stdin.readlines()
        #: system pipe
        pipe = "".join(pp)

    description = "Command-line tool for writing NeXus files" \
                  + " with NXSDataWriter"

    epilog = 'For more help:\n  nxsdata <sub-command> -h'
    parser = NXSArgParser(
        description=description, epilog=epilog,
        formatter_class=argparse.RawDescriptionHelpFormatter)
    parser.cmdrunners = [('openfile', OpenFile),
                         ('setdata', SetData),
                         ('openentry', OpenEntry),
                         ('record', Record),
                         ('closefile', CloseFile),
                         ('closeentry', CloseEntry)]
    runners = parser.createSubParsers()

    try:
        options = parser.parse_args()
    except ErrorException as e:
        sys.stderr.write("Error: %s\n" % str(e))
        sys.stderr.flush()
        parser.print_help()
        print("")
        sys.exit(255)

    if options.subparser is None:
        sys.stderr.write(
            "Error: %s\n" % str("too few arguments"))
        sys.stderr.flush()
        parser.print_help()
        print("")
        sys.exit(255)

    if options.subparser != 'servers':
        if not options.server:
            options.server = checkServer('NXSDataWriter')

        if not options.server:
            parser.subparsers[options.subparser].print_help()
            print("")
            sys.exit(255)

    #: command-line and pipe arguments
    parg = []
    if hasattr(options, "args"):
        parg = [options.args] if options.args else []
    if pipe:
        parg.append(pipe)
    options.args = parg

    result = runners[options.subparser].run(options)
    if result and str(result).strip():
        print(result)


if __name__ == "__main__":
    main()
